prov/gni/test/av.c

```diff
@@ -317,6 +361,422 @@ Test(av_full_table, lookup)
+				compare = (fi_addr_t *)
+					&simple_ep_names[j].gnix_addr;
+				if (addresses[i] == *compare) {
+					found_addresses[j] = true;
+				}
+			}
+		} else {
+			found_addresses[addresses[i]] = true;
+		}
+	}
+	for (i = 0; i < MT_ADDR_COUNT; i++) {
+		cr_assert_eq(found_addresses[i], true);
+	}
+
+	/* look up addresses */
+	for (i = 0; i < MT_ADDR_COUNT; i++) {
+		ret = pthread_create(&threads[i], NULL, lookup_single,
+				     (void *) &addresses[i]);
+		cr_assert_eq(ret, 0);
+	}
+
+	for (i = 0; i < MT_ADDR_COUNT; i++) {
+		ret = pthread_join(threads[i], (void **) &pret);
+		cr_assert_eq(ret, 0);
+		cr_assert_eq(pret, 0UL);
+	}
+
+	/* remove addresses */
+	for (i = 0; i < MT_ADDR_COUNT; i++) {
+		ret = pthread_create(&threads[i], NULL, remove_single,
+				     (void *) &addresses[i]);
+		cr_assert_eq(ret, 0);
+	}
+	for (i = 0; i < MT_ADDR_COUNT; i++) {
+		ret = pthread_join(threads[i], (void **) &pret);
+		cr_assert_eq(ret, 0);
+		cr_assert_eq(pret, 0UL);
+	}
+
+	ret = pthread_barrier_destroy(&mtbar);
+	cr_assert_eq(ret, 0);
+}
+
+Test(av_full_map, mt_simple)
+{
+	simple_mt_test();
+}
+
+Test(av_full_table, mt_simple)
+{
+	simple_mt_test();
+}
+
+#include "fi_atom.h"
+/* add a compare-and-swap */
+static inline int atomic_cas_weak(atomic_t *atomic, int *expected, int desired)
+{
+	ATOMIC_IS_INITIALIZED(atomic);
+	return atomic_compare_exchange_weak_explicit(&atomic->val,
+						     expected, desired,
+						     memory_order_seq_cst,
+						     memory_order_seq_cst);
+}
+
+static void *lookup_random(void *data)
+{
+	int n, ret;
+	fi_addr_t *addresses = ((fi_addr_t **) data)[0];
+	atomic_t *done = ((atomic_t **) data)[1];
+	struct gnix_ep_name found;
+	size_t addrlen = sizeof(struct gnix_ep_name);
+
+	srand(0);
+
+	ret = pthread_barrier_wait(&mtbar);
+	if ((ret != PTHREAD_BARRIER_SERIAL_THREAD) && (ret != 0)) {
+		pthread_exit((void *) 1UL);
+	}
+
+	while (!atomic_get(done)) {
+		n = rand()%MT_ADDR_COUNT;
+		(void) fi_av_lookup(av, addresses[n], &found, &addrlen);
+	}
+
+	pthread_exit(NULL);
+}
+
+static void continuous_lookup(void)
+{
+	int i, ret;
+	pthread_t thread;
+	fi_addr_t addresses[MT_ADDR_COUNT];
+	atomic_t done;
+	void *info[2];
+	const int iters = 17;
+
+	ret = pthread_barrier_init(&mtbar, NULL, 2);
+	cr_assert_eq(ret, 0);
+
+	init_av_lock();
+
+	atomic_initialize(&done, 0);
+
+	memset(addresses, 0, MT_ADDR_COUNT*sizeof(fi_addr_t));
+
+	info[0] = (void *) addresses;
+	info[1] = (void *) &done;
+
+	ret = pthread_create(&thread, NULL, lookup_random, info);
+	cr_assert_eq(ret, 0);
+
+	ret = pthread_barrier_wait(&mtbar);
+	cr_assert((ret == PTHREAD_BARRIER_SERIAL_THREAD) || (ret == 0));
+
+	for (i = 0; i < iters; i++) {
+		for (i = 0; i < MT_ADDR_COUNT; i++) {
+			ret = fi_av_insert(av, (void *) &simple_ep_names[i], 1,
+					   &addresses[i], 0, NULL);
+			cr_assert_eq(ret, 1);
+		}
+		for (i = 0; i < MT_ADDR_COUNT; i++) {
+			ret = fi_av_remove(av, &addresses[i], 1, 0);
+			cr_assert_eq(ret, FI_SUCCESS);
+		}
```

Comment:
> In multi-threaded tests, you often can't assume empty criteria until after the test is complete. Other threads may be in the process of removing the entries they are responsible for removing during your check to determine if the table/map is empty.

In this case, I believe one thread is just looking up while the other inserts all the addrs and then removes all of them.